{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0718094f-b9e0-413b-a601-a096133699b8",
   "metadata": {},
   "source": [
    "# <center><span style=\"color:Green\">Multi-modal Approach for Patient Benefit </span></center> \n",
    "\n",
    "<span style=\"color:Green\"> **Institute:** </span> Queen Mary University of London (Barts Cancer Institute)\n",
    "\n",
    "<span style=\"color:Green\"> **Date:** </span> 7th November 2024\n",
    "\n",
    "<span style=\"color:Green\"> **Created By:** </span> Chelala Team (QMUL)\n",
    "\n",
    "<span style=\"color:Green\"> **Email:** </span> c.chelala@qmul.ac.uk\n",
    "\n",
    "<span style=\"color:Green\"> **Status:** </span> OPTIMA Example\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1281900e-37d0-47ee-8402-02cb9633ad30",
   "metadata": {},
   "source": [
    "<span style=\"color:Green\"> Instructions to Create the Kernel</span><br>\n",
    "https://github.com/imcalledlewis/OPTIMA_Workshop_Nov_2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94adb574-0961-4421-932f-7427406e4bcb",
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# load libraries\n",
    "#library(TCGAbiolinks);\n",
    "#library(SummarizedExperiment);\n",
    "library(maftools);\n",
    "library(plotly);\n",
    "library(dplyr);\n",
    "library(org.Hs.eg.db);\n",
    "library(magrittr);\n",
    "library(dplyr);\n",
    "library(genefu);\n",
    "library(ggplot2);\n",
    "library(reshape2);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02870add-cf36-4e0d-bf81-4fa8dd2af04c",
   "metadata": {},
   "source": [
    "# <center><span style=\"color:Green\">1. The Cancer Genome Atlas (TCGA)</center>\n",
    "<br>\n",
    "<details> \n",
    "    <summary>1.1 What is the TCGA?</summary><br>\n",
    "<p>a.\tInitially set up in 2006 as a global initiative to molecularly characterise cancer and matched normal samples from major cancer types:</p>\n",
    "\n",
    "- As of Nov 2024: 86 projects; 69 primary sites; 44,736 cases across those sites.\n",
    "\n",
    "<p>b.   Idea is to have a well-annotated linked set of clinical and molecular data publicly available across these projects, processed in a standard way.</p>\n",
    "<p>c.\tAccessible via Genomic Data Commons Data Portal (https://portal.gdc.cancer.gov/) alongside other projects.</p>\n",
    "\n",
    "<center><img src=\"images/TCGA_front_page.png\" width=\"800\" /></center>\n",
    "    \n",
    "<p>d.\tTCGA generated over 2.5 PB of genomic, epigenomic, transcriptomic, proteomic and imaging data that is all available to the research community. </p>\n",
    "</details>\n",
    "<br>\n",
    "<details> \n",
    "    <summary>1.2 What data types are available for breast cancer?</summary><br>\n",
    "<p>a.\tClinical data: demographics, surgery information, tumour characteristics, treatment, co-morbidities, treatment response, ...</p> \n",
    "<p>b.\tMolecular data: genomic data, transcriptomic data, epigenetic data, ...</p>\n",
    "\n",
    "- 624 TB of data consisting of 61,140 data files from 1,097 unique donors.\n",
    "\n",
    "<center><img src=\"images/TCGA_exp_strategy.png\" width=\"800\" /></center>\n",
    "\n",
    "- <span style=\"color:green\">Open Access</span>/<span style=\"color:red\">Controlled Access</span>  \n",
    "  - While much of the data in the GDC is open access, many file types are controlled access. Processed data tends to be <span style=\"color:green\">open</span>, but raw data tends to be <span style=\"color:red\">controlled</span>. Access to controlled data is by application to the Data Access Committee.\n",
    "\n",
    "\n",
    "| Experimental Strategy\t| Cases\t| # Files | Size | Data format |\n",
    "| :-------------------- | :---: | :-----: | :--: | :---------- |\n",
    "| Diagnostic Slide (open) | 1,062 | 1,113 | 1.16 TB\t| <span style=\"color:green\">svs</span> (compressed image)|\n",
    "| Tissue Slide (open) | 1,092 | 1,976 | 479.3GB\t| <span style=\"color:green\">svs</span> (compressed image)|\n",
    "| RNA-seq (open/controlled)\t| 1,095\t| 1,231\t| 5.22 GB | <span style=\"color:green\">tsv</span> (gene counts)<br><span style=\"color:red\">bam, bedpe</span> | \n",
    "| WGS _(DNA-seq)_ (open/controlled) | 102 | 113 | 395.05 GB | <span style=\"color:green\">tsv, txt</span> (copy number variation – ASCAT)<br><span style=\"color:red\">bam</span> |\n",
    "| WXS _(DNA-seq)_ (open/controlled) | 970| 993 | 32.05 MB | <span style=\"color:green\">maf</span> (simple nucleotide variation)<br><span style=\"color:red\">vcf, bam</span> |\n",
    "| ATAC-seq (controlled)| 74 | 75 | 1.57 TB | <span style=\"color:red\">bam</span> |\n",
    "| Genotyping Array (open/controlled) | 1,098 | 14,329 | 214.72 GB | <span style=\"color:green\">txt</span> (genotype call) <br><span style=\"color:red\">tsv, cel</span> |\n",
    "| Methylation Array (open) | 1,097 | 3,714 | 26.93 GB | <span style=\"color:green\">txt, idat</span> (methylation call)|\n",
    "| miRNA-Seq (controlled/open) | 1,079 | 3,621 | 149.79 GB | <span style=\"color:green\">txt</span> (miRNA counts)<br><span style=\"color:red\">bam</span> |\n",
    "| Reverse Phase Protein Array (open) | 881 | 919 | 20.54 GB | <span style=\"color:green\">tsv</span> (protein level)|\n",
    "| Tissue Slide (open) | 1,093 | 1,977 | 479.62 GB | <span style=\"color:green\">svs</span> (compressed image)|\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb9491b4-2216-4caa-8ac3-b23f4e018b3f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <center><span style=\"color:Green\">2. Data Types</center>\n",
    "<br>\n",
    "<details> \n",
    "    <summary>2.1 What kind of data is available?</summary>\n",
    "\n",
    "<br>\n",
    "\n",
    "| Type\t| Examples |\n",
    "| :---- | :------- |\n",
    "| Demographics | age, sex, ethnicity, indices of deprivation |\n",
    "| Clinical | surgery: type <br> tumour: size, receptor status, invasive status |\n",
    "| Treatment | type, subtype, drug/drug class |\n",
    "| Co-morbidities | before / after malignancy |\n",
    "| Response to treatment | complete response, partial response, ... |\n",
    "| ... | ... |\n",
    "<p>\n",
    "Access to longitudinal clinical information from multiple sources is important to provide the most complete assessment of a patient’s journey over time. We will show an example of how this is important during this presentation.\n",
    "</p>\n",
    "</details>\n",
    "<br>\n",
    "<details> \n",
    "    <summary>2.2 What is genomics and how is it used in cancer research?</summary>\n",
    "<br>\n",
    "\n",
    "<p>a.\tInformation from sequencing DNA:\n",
    "\n",
    "- Variant data (germline and somatic)\n",
    "- Methylation data (epigenetic)\n",
    "- Copy number changes\n",
    "- Structural rearrangements</p>\n",
    "<br>\n",
    "    \n",
    "<p>b.\tGermline and somatic variants in breast cancer\n",
    "\n",
    "- **Germline** variants can indicate predisposition to cancer (e.g. _BRCA1/2_ risk factor)\n",
    "- **Somatic** variants influence tumour behaviour and response to treatments</p>\n",
    "</details>\n",
    "<br>\n",
    "<details> \n",
    "    <summary>2.3 What is transcriptomics and how is it used in cancer research?</summary>\n",
    "<br>\n",
    "\n",
    "<p>a.\tInformation from sequencing RNA:\n",
    "\n",
    "- Levels of gene expression\n",
    "- Levels of miRNA expression\n",
    "- Alternative splicing of RNA\n",
    "- Gene fusions</p>\n",
    "<br>\n",
    "\n",
    "<p>b.\tTranscriptomics in breast cancer\n",
    "\n",
    "- Treatments can affect gene expression (e.g. PI3K inhibitors, PARP inhibitors, monoclonal antibodies)\n",
    "- miRNAs can silence genes by binding to RNA to prevent translation\n",
    "- Different splice variants can produce different protein structures with different functions\n",
    "- Some fusions increase cancer aggressivity by switching on pathways otherwise switched off</p>\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c26bb359-96a8-4cc0-8349-bec8970e4960",
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### DEFINE VARIABLES USED IN CELL\n",
    "#proj.id <- \"TCGA-BRCA\";\n",
    "#data.cat <- \"Transcriptome Profiling\";\n",
    "#data.t <- \"Gene Expression Quantification\";\n",
    "#workflow.t <- \"STAR - Counts\";\n",
    "\n",
    "#### ACCESS CLINICAL DATA\n",
    "# define clinical data\n",
    "#clin.query <- GDCquery_clinic(proj.id, \"clinical\");\n",
    "\n",
    "# write clinical data to outfile\n",
    "#write.table(clin.query, file=paste(proj.id, \"_clinical.txt\", sep=\"\"), sep=\"\\t\", row.names=FALSE );\n",
    "\n",
    "#### ACCESS MOLECULAR DATA\n",
    "# this example uses expression data: read counts from RNAseq\n",
    "#exp.query <- GDCquery(\n",
    "#    project = proj.id,\n",
    "#    data.category = data.cat,\n",
    "#    data.type = data.t,\n",
    "#    workflow.type = workflow.t\n",
    "#);\n",
    "\n",
    "# download the expression matrix\n",
    "#GDCdownload(query = exp.query);\n",
    "#exp.data <- GDCprepare(\n",
    "#    query = exp.query,\n",
    "#    save = TRUE,\n",
    "#    save.filename = \"exp.rda\"\n",
    "#);\n",
    "\n",
    "#'exp.data' is a SummarizedExperiment object\n",
    "#expression_matrix <- assay(exp.data);\n",
    "\n",
    "# make gene symbols row names\n",
    "#rownames(expression_matrix) <- rowData(exp.data)$gene_name;\n",
    "\n",
    "# transpose the matrix: samples=columns and genes=rows\n",
    "#expression_matrix <- t(expression_matrix);\n",
    "\n",
    "# store all objects in an .RData file\n",
    "#save(expression_matrix, file = \"expression_matrix.RData\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1e34a59-808f-4784-9501-aa3a841d38c7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# <center><span style=\"color:Green\"> 3. Example </center></span>\n",
    "<b>Aim:</b> To examine breast cancer heterogeneity using clinical, genomic and transcriptomic features.<br>\n",
    "* <b><span style=\"color:Green\">Clinical Summary</span></b>\n",
    "  * Overview of the clinical characteristics of the cohort.\n",
    "* <b><span style=\"color:Green\">Genomic</span></b>\n",
    "  * Mutational landscape\n",
    "  * Stratification of patients based on mutational status\n",
    "  * Identify windows of therapeutic opportunity \n",
    "* <b><span style=\"color:Green\">Transcriptomic</span></b>\n",
    "  * Molecular classification of breast cancer\n",
    "  * Immunophenotyping of data using deconvolution methods\n",
    "* <b><span style=\"color:Green\">Whole slide imaging</span></b>\n",
    "  * Overlay our findings with whole slide image analysis to show how a multi-modal approach can enhance our understanding of breast cancer heterogeneity for patient benefit..\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4e117de-fa18-47a2-a1a8-8e672e8c259d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Green\"> 3.1 Clinical Summary</span>\n",
    "<b>Aim:</b> To visualise a summary of key clinical features available from <i>clinical_data.txt</i>.<br>\n",
    "* An <b><span style=\"color:Green\">interactive barplot</span></b> is generated, to visualise multiple features in relation to each other and identify potential trends in the data.<br><br>\n",
    "<sup>*The dropdown menu allows users to toggle between categories. </sup>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "effe19b0-32b4-4c0c-95ff-68e6de89bc29",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# load the clinical data\n",
    "clin.data <- read.table(\"clinical_data.txt\", header = TRUE, sep = \"\\t\")\n",
    "head(clin.data);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79144edc-8211-42b6-b8c2-a9b6023e4fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required libraries\n",
    "library(dplyr)\n",
    "library(plotly)\n",
    "\n",
    "# Define groups for Menopausal Status and Race\n",
    "age_groups <- c(\"<40\", \"40-60\", \">60\")\n",
    "race_groups <- c(\"White\", \"Black or African American\", \"Asian\", \"not available\")\n",
    "\n",
    "# Example data processing\n",
    "filtered_data <- clin.data %>%\n",
    "  mutate(race = factor(race, levels = race_groups)) %>%\n",
    "  mutate(age_group = case_when(\n",
    "    age < 40 ~ \"<40\",\n",
    "    age >= 40 & age <= 60 ~ \"40-60\",\n",
    "    age > 60 ~ \">60\"\n",
    "  )) %>%\n",
    "  mutate(age_group = factor(age_group, levels = age_groups))\n",
    "\n",
    "# Define color mappings for the categories\n",
    "colors_race <- c(\n",
    "  'White' = '#a0d0d0',\n",
    "  'Black or African American' = '#f0d0a0',\n",
    "  'Asian' = '#a0d0a0',\n",
    "  'not available' = '#d0d0d0'\n",
    ")\n",
    "\n",
    "# Create an empty plot\n",
    "plot <- plot_ly()\n",
    "\n",
    "# Add traces for Age Group by Race\n",
    "for (group in race_groups) {\n",
    "  group_data <- filtered_data %>%\n",
    "    filter(race == group)\n",
    "  \n",
    "  total_counts <- filtered_data %>%\n",
    "    group_by(age_group) %>%\n",
    "    summarize(total_count = n(), .groups = 'drop')\n",
    "  \n",
    "  summary_data <- group_data %>%\n",
    "    group_by(age_group) %>%\n",
    "    summarize(count = n(), .groups = 'drop') %>%\n",
    "    left_join(total_counts, by = \"age_group\") %>%\n",
    "    mutate(proportion = (count / total_count) * 100)  # Proportion as a percentage\n",
    "  \n",
    "  plot <- plot %>%\n",
    "    add_trace(\n",
    "      data = summary_data,\n",
    "      x = ~proportion,\n",
    "      y = ~age_group,\n",
    "      type = 'bar',\n",
    "      name = group,\n",
    "      text = ~paste(\"Count: \", count, \"<br>Proportion: \", round(proportion, 1), \"%\"),\n",
    "      hoverinfo = \"text\",\n",
    "      textposition = 'none',\n",
    "      orientation = 'h',\n",
    "      marker = list(color = colors_race[[group]])\n",
    "    )\n",
    "}\n",
    "\n",
    "# Update plot layout to display Age Group by Race\n",
    "plot <- plot %>%\n",
    "  layout(\n",
    "    barmode = \"stack\",\n",
    "    title = \"Age Group by Race\",  # Title for Age Group by Race view\n",
    "    xaxis = list(title = \"Percentage of Donors\"),\n",
    "    yaxis = list(title = \"Age Group\")  # Y-axis title for Age Group\n",
    "  )\n",
    "\n",
    "# Display the plot\n",
    "plot\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cf599e0-9ce5-424a-b1b7-dc4bbaf186f9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Green\"> 3.2 Genomics </span>\n",
    "\n",
    "<b>Aim:</b> To generate a mutational landscape of the cohort and determine how to use this data to identify patients with therapeutically actionable targets.<br>\n",
    "<b>\n",
    "Description</b>: Use information provided in tcga.maf and clinical_data.txt to generate summary and focussed variant plots and stratify the cohort based on pharmacogenomic potential.<br>\n",
    "* A <b><span style=\"color:Green\">summary plot</span></b> is generated, displaying the range of variant classifications, variant types and base substitution profiles as bar plots and/or box plots as well as the number of variants in each sample is available as a stacked bar plot and a summary of the top mutated genes for the cohort.\n",
    "* The <span style=\"color:Green\"><b>oncoplot</span></b> provides a mutational landscape of the top 25 mutated genes in the cohort and allows for inclusion of features from the clinical file.\n",
    "* The <span style=\"color:Green\"><b>lollipop plot</span></b> shows the amino acid changes within a defined gene (PIK3CA). These plots display the observed mutation distribution and protein domains, which are labelled for the selected gene. A summary of the observed mutation rate for the gene is also provided.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ec9be6c-ac97-44f9-8b4b-968a86a364f0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### LOAD LIBRARIES\n",
    "library(\"maftools\");\n",
    "\n",
    "\n",
    "#### DEFINE VARIABLES AND PARAMETERS TO BE USED IN THE CELL\n",
    "outDir <- getwd();\n",
    "\n",
    "aa_mut <- \"aaChange\"; # column relating to amino acid change\n",
    "flag_genes <- \"TTN\"; # genes to remove\n",
    "top.mutated <- 25; # no. genes to plot\n",
    "\n",
    "# set plot size for the notebook and white background\n",
    "options(repr.plot.width = 12, repr.plot.height = 10);\n",
    "\n",
    "\n",
    "#### READ IN DATA FILES\n",
    "# name of variant data file\n",
    "maf.file <- paste0(\"tcga.maf\");\n",
    "\n",
    "# rename 'submitter_id' to 'Tumor_Sample_Barcode'\n",
    "colnames(clin.data)[colnames(clin.data) == \"submitter_id\"] <- \"Tumor_Sample_Barcode\"\n",
    "\n",
    "# Reorder the clin data subtypes in sensible plotting order\n",
    "clin.data$Subtype <- factor(clin.data$Subtype, levels = c(\"Normal\", \"LumA\", \"LumB\", \"Her2\", \"Basal\"))\n",
    "\n",
    "# read the maf file and the clinical data file\n",
    "var.data <- read.maf(maf=maf.file, clinicalData=clin.data)\n",
    "\n",
    "# filter out flagged genes\n",
    "var.data <- filterMaf(maf=var.data, genes=flag_genes);\n",
    "\n",
    "#### IDENTIFY TOP MUTATED GENES\n",
    "# Write top mutated genes to outfile\n",
    "gene.summary <- getGeneSummary(var.data );\n",
    "gene.top200 <- gene.summary[order(-gene.summary$MutatedSamples),]$Hugo_Symbol[1:200];\n",
    "#write.table(gene.top200, file=paste0(outDir, \"/top_genes.txt\"), quote = F, row.names = F, col.names = F);\n",
    "gene <- gene.top200[2] # Define gene of interest for lollipop plot and druggability analysis\n",
    "\n",
    "#### PLOT SUMMARISED DATA\n",
    "# Set larger size and white background\n",
    "png(paste0(outDir, \"/TCGA_variant_summary_plot.png\"), width=12, height=10, units=\"in\", res=300, bg=\"white\")\n",
    "plotmafSummary(\n",
    "    maf=var.data,\n",
    "    rmOutlier=TRUE,\n",
    "    addStat='median',\n",
    "    dashboard=TRUE,\n",
    "    titvRaw=FALSE,\n",
    "    titleSize=1.5,\n",
    "    fs=1\n",
    ")\n",
    "dev.off()\n",
    "\n",
    "# Display the plot inline in the notebook\n",
    "par(bg = \"white\") # Set the background for the notebook plots\n",
    "plotmafSummary(\n",
    "    maf=var.data,\n",
    "    rmOutlier=TRUE,\n",
    "    addStat='median',\n",
    "    dashboard=TRUE,\n",
    "    titvRaw=FALSE,\n",
    "    titleSize=1.5,\n",
    "    fs=1\n",
    ")\n",
    "\n",
    "\n",
    "#### PLOT ONCOPLOT\n",
    "# oncoplot of top 25 mutated genes\n",
    "for (t in top.mutated) {\n",
    "    png(paste0(outDir, \"/TCGA_oncoplot_\", t, \"genes.png\"), width=12, height=10, units=\"in\", res=300, bg=\"white\")\n",
    "    oncoplot(\n",
    "        maf=var.data,\n",
    "        draw_titv=TRUE,\n",
    "        top=t,\n",
    "        fontSize=1,\n",
    "        clinicalFeatures = c(\"er\"),\n",
    "        sortByAnnotation=TRUE,\n",
    "        titleFontSize=1.5,\n",
    "        annotationFontSize=1.2,\n",
    "        legendFontSize=1.2\n",
    "    )\n",
    "    dev.off()\n",
    "\n",
    "    # Display oncoplot inline in the notebook\n",
    "    par(bg = \"white\") # Set white background\n",
    "    oncoplot(\n",
    "        maf=var.data,\n",
    "        draw_titv=FALSE,\n",
    "        top=t,\n",
    "        fontSize=1,\n",
    "        clinicalFeatures = c(\"er\"),\n",
    "        sortByAnnotation=TRUE,\n",
    "        titleFontSize=1.5,\n",
    "        annotationFontSize=1.2,\n",
    "        legendFontSize=1.2 \n",
    "    )\n",
    "}\n",
    "\n",
    "\n",
    "#### PLOT LOLLIPOP PLOT\n",
    "if (gene %in% gene.summary$Hugo_Symbol) {\n",
    "    pngfile <- paste0(outDir, \"/TCGA_lollipop_\", gene, \".png\")\n",
    "    png(pngfile, width=14, height=10, units=\"in\", res=300, bg=\"white\")\n",
    "\n",
    "    lollipopPlot(\n",
    "        maf = var.data, \n",
    "        gene = gene, \n",
    "        AACol = aa_mut, \n",
    "        showMutationRate = TRUE, \n",
    "        domainLabelSize = 0.8, \n",
    "        labelPos=c(542,545,1047),\n",
    "        pointSize = 0.9,\n",
    "        defaultYaxis = FALSE\n",
    "    )\n",
    "    dev.off()\n",
    "\n",
    "    # Display the lollipop plot inline in the notebook\n",
    "    par(bg = \"white\") # Set white background\n",
    "    lollipopPlot(\n",
    "        maf = var.data, \n",
    "        gene = gene, \n",
    "        AACol = aa_mut, \n",
    "        showMutationRate = TRUE, \n",
    "        domainLabelSize = 0.8, \n",
    "        labelPos=c(542,545,1047),\n",
    "        pointSize = 0.9,\n",
    "        defaultYaxis = FALSE\n",
    "    )\n",
    "} else {\n",
    "    cat(gene, \" not in gene.summary\\n\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f51d61-6591-4d51-a091-4ebc97265593",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Green\"> 3.3 Pharmacogenomic Implications </span>\n",
    "\n",
    "<b>Aim:</b> To identify therapeutically actionable targets in the cohort based on variant data.<br>\n",
    "<b>\n",
    "\n",
    "* An <b><span style=\"color:Green\">alluvial plot</span></b> is generated, displaying the targets alongside drug response."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab67cb6c-8644-4a7e-a817-6d32dca56fea",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### LOAD LIBRARIES\n",
    "library(ggplot2);\n",
    "library(ggalluvial);\n",
    "library(ggtext);\n",
    "\n",
    "#### READ IN DATA FILES\n",
    "# read pharmacogenomic table\n",
    "alluvial.data <- read.table(\"actionability.txt\", sep=\"\\t\", header=T);\n",
    "\n",
    "#### PREPARE DATA INPUT AND FORMAT\n",
    "# define colour palette for treatment response\n",
    "palette <- c(Resistant = \"#ff0000ff\", Responsive = \"#00a000ff\");\n",
    "\n",
    "# generate input\n",
    "bio_drug.uniq <- unique(alluvial.data[ , c(\"Gene\",\"Classification\",\"Response\")]);\n",
    "\n",
    "# identify unique entries\n",
    "alluvial.uniq <- unique(bio_drug.uniq);\n",
    "\n",
    "# add NA column to be filled by patient count - determines the \"thickness\" of each band\n",
    "dataset <- cbind(bio_drug.uniq, weight=NA);\n",
    "\n",
    "# go through unique rows in dataset and count how many patients from alluvial.uniq\n",
    "for( i in 1:length(dataset$weight) ){\n",
    "  # identify the biomarker count\n",
    "  dataset$weight[i] <- sum(\n",
    "    alluvial.uniq$Gene == dataset$Gene[i] &\n",
    "      alluvial.uniq$Classification == dataset$Classification[i] &\n",
    "      alluvial.uniq$Response == dataset$Response[i]\n",
    "  )\n",
    "}\n",
    "\n",
    "dataset$Response <- factor(\n",
    "  dataset$Response, levels = c(\"Resistant\", \"Responsive\")\n",
    ");\n",
    "dataset$Response <- droplevels(dataset$Response);\n",
    "\n",
    "#### GENERATE ALLUVIAL PLOT\n",
    "p <- (\n",
    "  ggplot(dataset,aes(y=weight, axis1=Gene, axis2=Classification)) +\n",
    "    geom_alluvium(aes(fill=Response), width=1/2, knot.pos=0.2, reverse=TRUE) +\n",
    "    geom_stratum(width=1/2, fill=\"white\", color=\"black\", alpha=0.5) +\n",
    "    scale_fill_manual(values = palette) +\n",
    "    geom_richtext(stat=\"stratum\", aes(label=after_stat(stratum)), size=2.7, label.size=NA,fill=NA) +\n",
    "    scale_x_discrete(limits=c(\"Gene\", \"Drug Classification\")) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      axis.text.x=element_text(size=12, color=\"black\"),\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.y=element_blank(),\n",
    "      axis.ticks.y=element_blank(),\n",
    "      axis.ticks.x=element_blank(),\n",
    "      panel.grid.major=element_blank(),\n",
    "      panel.grid.minor=element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.background = element_blank(),\n",
    "      plot.title=element_text(hjust=0.5)\n",
    "    ) +\n",
    "    theme(legend.position=\"bottom\") +\n",
    "    ggtitle(\"Clinically-actionable targets\")\n",
    ")\n",
    "\n",
    "print(p);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41477dec-731f-4368-bf8b-71d5dbd20f64",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### READ IN DATA FILES\n",
    "clin.data <- read.table(\"clinical_data.txt\", header = TRUE, sep = \"\\t\");\n",
    "\n",
    "# focus on expression profiles of ER- tumours\n",
    "clin.data <- clin.data[ clin.data$Subtype==\"Basal\"| clin.data$Subtype==\"Her2\" , ];\n",
    "\n",
    "# survival status\n",
    "clin.data$surv.stat <- ifelse(clin.data$surv.stat == \"dead\", 1, 0)\n",
    "\n",
    "# set plotting parameters for a white background\n",
    "par(bg = \"white\")\n",
    "\n",
    "#### PLOT SURVIVAL\n",
    "# define gene(s) of interest\n",
    "genes_list <- c(\"CDH1\", \"BRCA1\", \"BRCA2\", \"ERBB2\")\n",
    "\n",
    "mafSurvival(\n",
    "    maf = var.data,\n",
    "    genes = genes_list,\n",
    "    clinicalData = clin.data,\n",
    "    time = \"surv.time\",\n",
    "    Status = \"surv.stat\",\n",
    "    groupNames = c(\"Mutant\", \"WT\"),\n",
    "    showConfInt = TRUE,\n",
    "    addInfo = TRUE,\n",
    "    col = c(\"maroon\", \"royalblue\"),\n",
    "    isTCGA = TRUE,\n",
    "    textSize = 12\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a094e53-5a50-46c7-ae95-06850fd76e30",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Green\"> 3.4 Transcriptomics </span>\n",
    "\n",
    "<b>Aim:</b> To use clinical and transcriptomic data to look at cancer heterogeneity.<br>\n",
    "<b>Description:</b>Use information from <i>clinical_data.txt</i> and <i>expression_matrix.RData</i> to conduct high level determinations of molecular subtype and inferrences of immune populations.<br>\n",
    "\n",
    "* The <b><span style=\"color:Green\">PAM50 Subtype</span></b> predictor is applied to the expression matrix to classify breast cancer patients into molecular subtypes — basal-like, luminal A, luminal B, Her2-enriched, and normal-like.\n",
    "* <b><span style=\"color:Green\">xCell </span></b> is applied to analyse cellular heterogeneity and reconstruct the cellular composition of tissues. This method uses the gene expression matrix as input to infer the abundance of up to 64 cell types.\n",
    "\n",
    "<br><p><sup>*Data is derived from the TCGA BRCA cohort, but similar workflows could be applied to other transcriptomic datasets e.g., OPTIMA datasets: RNAseq from the PALOMA-3 trial (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE128500).</sup></p>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ad453e12-3c81-41ad-ab34-26e861eca7f6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘genefu’ was built under R version 4.3.2”\n",
      "Loading required package: survcomp\n",
      "\n",
      "Warning message:\n",
      "“package ‘survcomp’ was built under R version 4.3.2”\n",
      "Loading required package: survival\n",
      "\n",
      "Warning message:\n",
      "“package ‘survival’ was built under R version 4.3.3”\n",
      "Loading required package: prodlim\n",
      "\n",
      "Warning message:\n",
      "“package ‘prodlim’ was built under R version 4.3.3”\n",
      "Loading required package: biomaRt\n",
      "\n",
      "Warning message:\n",
      "“package ‘biomaRt’ was built under R version 4.3.2”\n",
      "Loading required package: iC10\n",
      "\n",
      "Warning message:\n",
      "“package ‘iC10’ was built under R version 4.3.3”\n",
      "Loading required package: AIMS\n",
      "\n",
      "Warning message:\n",
      "“package ‘AIMS’ was built under R version 4.3.2”\n",
      "Loading required package: e1071\n",
      "\n",
      "Warning message:\n",
      "“package ‘e1071’ was built under R version 4.3.3”\n",
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.3.3”\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.3.2”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### LOAD LIBRARIES AND RDATA FILES\n",
    "library(genefu);\n",
    "library(org.Hs.eg.db);\n",
    "#Load expression data\n",
    "load(\"expression_matrix.RData\");\n",
    "\n",
    "#### FORMAT DATA\n",
    "# get PAM50 model\n",
    "data(pam50.robust)\n",
    "# load gene symbols\n",
    "my.symbols <- colnames(expression_matrix) \n",
    "\n",
    "# map to Entrez IDs using AnnotationDbi::select()\n",
    "annot <- AnnotationDbi::select(\n",
    "    org.Hs.eg.db, keys = my.symbols,\n",
    "    columns = c(\"ENTREZID\", \"SYMBOL\"),\n",
    "    keytype = \"SYMBOL\"\n",
    ");\n",
    "\n",
    "colnames(annot) <- c(\"SYMBOL\", \"EntrezGene.ID\");\n",
    "\n",
    "# filter to keep one-to-one mapping\n",
    "annot <- distinct(annot, SYMBOL, .keep_all = TRUE);\n",
    "\n",
    "# add the \"probe\" column and duplicate the SYMBOL column\n",
    "annot$probe <- annot$SYMBOL #genefu requires a probe column\n",
    "\n",
    "#### DEFINE CLASSIFIER AND APPLY TO EACH SAMPLE\n",
    "# apply PAM50 classifier \n",
    "SubtypePredictions <- molecular.subtyping(\n",
    "    sbt.model = \"pam50\",\n",
    "    data = expression_matrix,\n",
    "    annot = annot,\n",
    "    do.mapping = TRUE,\n",
    "    verbose = TRUE\n",
    ");\n",
    "\n",
    "#### GENERATE PIECHART\n",
    "# calculate the frequency of each subtype\n",
    "subtype_counts <- table(SubtypePredictions$subtype);\n",
    "\n",
    "# create a pie chart\n",
    "par(bg = \"white\") # set the background color to white\n",
    "pie(\n",
    "    subtype_counts, \n",
    "    main = \"Distribution of PAM50 Subtypes\",\n",
    "    col = rainbow(length(subtype_counts)), \n",
    "    labels = paste0(names(subtype_counts), \" (\", round(subtype_counts/sum(subtype_counts) * 100, 1), \"%)\") \n",
    ");\n",
    "\n",
    "# add legend\n",
    "legend(\n",
    "    \"topright\",\n",
    "    legend = names(subtype_counts),\n",
    "    fill = rainbow(length(subtype_counts)),\n",
    "    cex = 0.8\n",
    ") \n",
    "#### LOAD LIBRARIES\n",
    "par(bg = \"transparent\") # reset the background colour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6711ed1-913e-4e22-8537-a8c7b4916f50",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### LOAD LIBRARIES AND RDATA FILES\n",
    "library(xCell);\n",
    "load(\"expression_matrix.RData\");\n",
    "\n",
    "#### RUN XCELL ANALYSIS\n",
    "results <- xCellAnalysis(t(expression_matrix), parallel.sz = 1);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ca0bab2-6292-48b3-b226-c65cc1234102",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### REFORMAT DATA AND GROUP FOR GGPLOTS\n",
    "# convert the thresholded predictions to data frame\n",
    "subtype_df <- as.data.frame(SubtypePredictions$subtype.crisp)\n",
    "subtype_df$SampleID <- rownames(subtype_df)\n",
    "\n",
    "# melt the DataFrame to long format to get samples with their subtypes\n",
    "subtype_long <- melt(subtype_df, id.vars = \"SampleID\")\n",
    "\n",
    "# filter to include only the rows where the subtype is 1 (indicating presence) and rename columns\n",
    "subtype_result <- subtype_long[subtype_long$value == 1, c(\"SampleID\", \"variable\")]\n",
    "colnames(subtype_result) <- c(\"SampleID\", \"Subtype\")\n",
    "\n",
    "# create a named vector for easy access\n",
    "subtype_mapping <- setNames(subtype_result$Subtype, subtype_result$SampleID);\n",
    "\n",
    "# get the column names (SampleIDs) from the results matrix\n",
    "results_colnames <- colnames(results);\n",
    "\n",
    "## group the results by subtype\n",
    "# create an empty list to hold the grouped results\n",
    "grouped_results <- list();\n",
    "\n",
    "# loop through unique subtypes and subset the matrix to build grouped results\n",
    "for (subtype in unique(subtype_result$Subtype)) {\n",
    "  # get the SampleIDs for the current subtype\n",
    "  sample_ids <- names(subtype_mapping[subtype_mapping == subtype])\n",
    "  \n",
    "  # subset the results matrix for these SampleIDs\n",
    "  if (length(sample_ids) > 0) {\n",
    "    grouped_results[[subtype]] <- results[, colnames(results) %in% sample_ids, drop = FALSE]\n",
    "  }\n",
    "}\n",
    "\n",
    "names(grouped_results) <- unique(subtype_result$Subtype);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e620a31-f79f-4ccf-bf4d-572b8080dae7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### LOAD LIBRARIES\n",
    "library(ggplot2);\n",
    "library(patchwork);\n",
    "\n",
    "#### DEFINE PARAMETERS\n",
    "# create a larger plotting area\n",
    "options(repr.plot.width = 20, repr.plot.height = 15);\n",
    "\n",
    "# Calculate mean proportions\n",
    "mean_proportions_list <- lapply(grouped_results, function(mat) {\n",
    "  rowMeans(mat, na.rm = TRUE)\n",
    "})\n",
    "\n",
    "# set maximum y-axis limit based on the max value from all groups\n",
    "max_y_limit <- max(sapply(mean_proportions_list, max, na.rm = TRUE));\n",
    "\n",
    "# define subtypes to plot\n",
    "first_four_subtypes <- names(mean_proportions_list)[1:4];\n",
    "num_subtypes <- length(first_four_subtypes);\n",
    "\n",
    "# initialize an empty list to store plots\n",
    "plot_list <- vector(\"list\", num_subtypes);\n",
    "score_plots <- vector(\"list\", num_subtypes);\n",
    "\n",
    "#### GENERATE PLOT\n",
    "# loop through each subtype and create a plot\n",
    "for (i in 1:num_subtypes) {\n",
    "  subtype <- first_four_subtypes[i];\n",
    "  mean_proportions <- mean_proportions_list[[subtype]];\n",
    "  \n",
    "  # create a data frame for ggplot\n",
    "  df <- data.frame(\n",
    "    CellType = rownames(grouped_results[[subtype]]),\n",
    "    Proportion = mean_proportions\n",
    "  );\n",
    "\n",
    "  # move the Scores to the end of the barplots\n",
    "  df$CellType <- factor(df$CellType, levels=df$CellType)\n",
    "   \n",
    "  # create a bar plot and save it to the list\n",
    "  plot_list[[i]] <- ggplot(df, aes(x = CellType, y = Proportion, fill = CellType)) +\n",
    "    geom_bar(stat = \"identity\") +\n",
    "    scale_fill_manual(values = c(rainbow(length(mean_proportions)-3), rep(\"darkgreen\", 3))) +\n",
    "    labs(title = paste(\"Average Cell Type Proportions for\", subtype), \n",
    "         y = \"Average Proportion\") +\n",
    "    theme_minimal() + \n",
    "    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),\n",
    "          plot.title = element_text(hjust = 0.5)) +\n",
    "    ylim(0, max_y_limit) +\n",
    "    guides(fill = \"none\")\n",
    "    \n",
    "  df <- df %>% filter(CellType %in% c(\"ImmuneScore\", \"StromaScore\", \"MicroenvironmentScore\"))\n",
    "    \n",
    "  score_plots[[i]] <- ggplot(df, aes(x = CellType, y = Proportion, fill = CellType)) +\n",
    "    geom_bar(stat = \"identity\") +\n",
    "    scale_fill_manual(values = rep(\"darkgreen\", 3)) +\n",
    "    labs(title = paste(\"Average Cell Scores for\", subtype), \n",
    "         y = \"Scores\") +\n",
    "    theme_minimal() + \n",
    "    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),\n",
    "          plot.title = element_text(hjust = 0.5)) +\n",
    "    ylim(0, max_y_limit) +\n",
    "    guides(fill = \"none\")\n",
    "     \n",
    "    \n",
    "}\n",
    "\n",
    "# combine the plots using patchwork\n",
    "combined_plot <- (plot_list[[1]] | plot_list[[2]]) / (plot_list[[3]] | plot_list[[4]]);\n",
    "combined_scores <- (score_plots[[1]] | score_plots[[2]] | score_plots[[3]] | score_plots[[4]])\n",
    "\n",
    "# print the combined plot\n",
    "print(combined_plot);\n",
    "\n",
    "# print the scores only as lower row\n",
    "options(repr.plot.width = 20, repr.plot.height = 5);\n",
    "print(combined_scores);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BCI_demo",
   "language": "R",
   "name": "bci_demo"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
